File: Prolog/decodificador.pl
:- initialization(main).

decodificar(Codigo, Resultado) :-
    number_string(Codigo, CodigoTexto),
    string_length(CodigoTexto, Largo),
    Largo =:= 8,                                   % debe tener exactamente 8 digitos

    sub_string(CodigoTexto, 0, 3, _, TextoPeriodo), % digitos 1 a 3
    number_string(Prefijo, TextoPeriodo),
    obtener_periodo(Prefijo, Periodo),              % falla si esta fuera de 262-292

    sub_string(CodigoTexto, 3, 2, _, TextoCategoria),% digitos 4 y 5
    number_string(NumCategoria, TextoCategoria),
    obtener_categoria(NumCategoria, Categoria),

    sub_string(CodigoTexto, 5, 3, _, TextoConsec),  % digitos 6 a 8
    number_string(Consecutivo, TextoConsec),        % number_string ya quita los ceros

    obtener_paridad(Consecutivo, Paridad),

    format(atom(Resultado), "~w ~w num~w ~w", [Periodo, Categoria, Consecutivo, Paridad]).

% obtener_periodo(+Prefijo, -Periodo)

obtener_periodo(Prefijo, Periodo) :-
    Prefijo >= 262,
    Prefijo =< 292,
    Anio is 2000 + (Prefijo // 10),
    Semestre is Prefijo mod 10,
    format(atom(Periodo), "~w-~w", [Anio, Semestre]).


% divisores_propios(+Numero, -Divisores)

divisores_propios(Numero, Divisores) :-
    findall(D,
            ( between(1, Numero, D),
              D < Numero,
              0 is Numero mod D
            ),
            Divisores).

suma_divisores(Numero, Suma) :-
    divisores_propios(Numero, Divisores),
    sum_list(Divisores, Suma).

% obtener_categoria(+Numero, -Categoria)

obtener_categoria(Numero, "Administrativa") :-
    suma_divisores(Numero, Suma),
    Suma > Numero,
    !.
obtener_categoria(Numero, "Ingeniería") :-
    suma_divisores(Numero, Suma),
    Suma =:= Numero,
    !.
obtener_categoria(_Numero, "Humanidades").

% obtener_paridad(+Numero, -Paridad)

obtener_paridad(Numero, "even") :-
    0 is Numero mod 2,
    !.
obtener_paridad(_Numero, "odd").

% generar_codigos(+Periodo, +Categoria, -Codigos)
% Consulta en modo "generar"

generar_codigos(Periodo, Categoria, Codigos) :-
    Desde is Periodo * 100000,
    Hasta is Periodo * 100000 + 99999,
    findall(Codigo,
            ( between(Desde, Hasta, Codigo),
              decodificar(Codigo, Resultado),
              sub_string(Resultado, _, _, _, Categoria)
            ),
            Codigos).

% main/0: Pruebas del programa

main :-
    probar(26276002),   % ejemplo del enunciado
    probar(27110123),   % otro ejemplo
    probar(26228002),   % numero perfecto (28)
    probar(26212003),   % numero abundante (12)
    ( decodificar(1234567, _)   % 7 digitos: debe fallar
    -> writeln('ERROR: no debia funcionar con 7 digitos')
    ;  writeln('OK: fallo correctamente con codigo de 7 digitos')
    ),
    ( decodificar(30012345, _)  % periodo fuera de rango: debe fallar
    -> writeln('ERROR: no debia funcionar con periodo fuera de rango')
    ;  writeln('OK: fallo correctamente con periodo fuera de rango')
    ),
    generar_codigos(292, "Ingeniería", CodigosIngenieria),
    length(CodigosIngenieria, Cantidad),
    format("Codigos 2029-2 Ingenieria encontrados: ~w~n", [Cantidad]),
    writeln(CodigosIngenieria),
    halt.

probar(Codigo) :-
    ( decodificar(Codigo, Resultado)
    -> writeln(Resultado)
    ;  writeln('El codigo no es valido')
    ).

File: Haskell/decodificador.hs
module Main where

-- Funcion principal: junta las 4 partes del resultado en un solo texto.
-- Usamos "where" para separar cada paso del calculo.
decodificar :: Int -> String
decodificar codigo =
    periodoTexto ++ " " ++ categoriaTexto ++ " num" ++ show consecutivo ++ " " ++ paridadTexto
  where
    codigoTexto    = show codigo                       -- pasamos el numero a texto para poder cortarlo
    prefijo        = read (take 3 codigoTexto) :: Int          -- digitos 1 a 3
    digitosCat     = read (take 2 (drop 3 codigoTexto)) :: Int -- digitos 4 y 5
    consecutivo    = read (drop 5 codigoTexto) :: Int          -- digitos 6 a 8 (read ya quita los ceros)
    periodoTexto   = obtenerPeriodo prefijo
    categoriaTexto = obtenerCategoria digitosCat
    paridadTexto   = if consecutivo `mod` 2 == 0 then "even" else "odd"

-- Convierte el prefijo (por ejemplo 262 o 271) al formato "Año-Semestre".
-- Usa guardas para validar el rango pedido (262 a 292).
obtenerPeriodo :: Int -> String
obtenerPeriodo prefijo
    | prefijo < 262 || prefijo > 292 = error "El periodo esta fuera del rango 262-292"
    | otherwise                      = show anio ++ "-" ++ show semestre
  where
    anio     = 2000 + (prefijo `div` 10)   -- los dos primeros digitos del prefijo
    semestre = prefijo `mod` 10            -- el ultimo digito del prefijo

-- Calcula la suma de los divisores propios de n (todos menos el mismo n).
-- Es una comprension de listas sencilla: recorremos del 1 a n-1 y nos
-- quedamos con los que dividen a n exactamente (resto 0).
sumaDivisores :: Int -> Int
sumaDivisores n = sum [d | d <- [1 .. n - 1], n `mod` d == 0]

-- Clasifica el numero de dos digitos segun la suma de sus divisores
-- (numero de Nicomaco: abundante, perfecto o deficiente).
obtenerCategoria :: Int -> String
obtenerCategoria n
    | suma > n  = "Administrativa"   -- abundante
    | suma == n = "Ingeniería"       -- perfecto
    | otherwise = "Humanidades"      -- deficiente
  where
    suma = sumaDivisores n

-- main solo se usa para probar el programa con algunos ejemplos.
main :: IO ()
main = do
    putStrLn (decodificar 26276002)  -- ejemplo del enunciado
    putStrLn (decodificar 27110123)  -- otro ejemplo de prueba
    putStrLn (decodificar 26228002)  -- ejemplo con numero perfecto (28)
    putStrLn (decodificar 26212003)  -- ejemplo con numero abundante (12)
